### DISCUSSION TITLE
Can I refer to the other prop inside a prop?

### DISCUSSION THREAD

**Discussion by sheremet-va** [vuejs/language-tools#143] [Q&A] [answered]
Can I do something like this using Volar, and get the auto-completion and type checking support?

```
interface MyObject {
   [key: string]: any
}

defineComponent({
  props: {
    value: Object as PropType<T extends MyObject>,
    key: String as PropType<keyof T>
  }
})
```

**Comment by johnsoncodehk** [comment]
Hi @sheremet-va, as I know `defineComponent` do not support this, but you can use a wrapper function.

**Update:** this way do not support `<template>`, use the type-only way for replacement: https://github.com/johnsoncodehk/volar/discussions/143#discussioncomment-634351

```ts
// child.vue
import { defineComponent, PropType } from 'vue'

export default defineChild();

interface MyObject {
  [key: string]: any
}

export function defineChild<T extends MyObject>() {
  return defineComponent({
    props: {
      value: Object as PropType<T>,
      key: String as PropType<keyof T>
    },
  })
}
```

```ts
// parent.vue
import { defineComponent } from 'vue'
import { defineChild } from './child.vue'

export default defineComponent({
  components: {
    Child: defineChild<{ foo: string, bar: number }>()
  }
})
```

**Reply by sheremet-va** [reply]
Looks interesting. I'm gonna use it! 

But it would be cool if we could use some kind of type enhancement like with `__VLS_slots` (I know you will replace them in the future tho).

**Reply by sheremet-va** [reply]
Also it doesn't work if I want to use the child component multiple times with different objects. :(

**Reply by johnsoncodehk** [reply]
@sheremet-va I avoid to create more `__VLS_...`, because this is a volar-specific thing.

**Comment by johnsoncodehk** [comment, marked as answer]
Also has a type-only way, it should be support reuse the child component. A challenge is how to define `DefineComponent<...>` correctly.

```ts
// child.vue
import { defineComponent, PropType } from 'vue'

interface MyObject {
  [key: string]: any
}

export default defineComponent({
  props: {
    value: Object as PropType<MyObject>,
    key: String as PropType<keyof MyObject>
  },
}) as ChildType<MyObject>;

export type ChildType<T extends MyObject> = DefineComponent<{
  value: PropType<T>,
  key: PropType<keyof T>,
}>;
```

```ts
// parent.vue
import { defineComponent } from 'vue'
import Child, { ChildType } from './child.vue'

export default defineComponent({
  components: {
    Child: Child as ChildType<{ foo: string, bar: number }>
  }
})
```

**Reply by sheremet-va** [reply]
to reuse it you suggest defining component multiple times? like
```ts
export default defineComponent({
  components: {
     ChildOne: Child as ChildType<{ foo: string }>,
     ChildeTwo: Child as ChildType<{ bar: number}>
  }
})
```

**Reply by johnsoncodehk** [reply]
Sorry I misunderstood your "reuse" mean, yes you have to do it. :P

Your request is generic type component, I think it's worth to open a discussion in https://github.com/vuejs/rfcs/discussions.

**Comment by sheremet-va** [comment]
Hello again! I found myself using new script setup sugar (which is still an rfc, I know) a lot, and I was wondering if I can get the same type checking and autocomplete using script setup syntax?

Even with some hacks?

**Reply by johnsoncodehk** [reply]
Yes it's similar.

```vue
// child.vue
<script lang="ts" setup>
defineProps({
  value: Object as PropType<MyObject>,
  key: String as PropType<keyof MyObject>
});
</script>

<script lang="ts">
import type { DefineComponent, PropType } from 'vue'
import { defineProps } from 'vue'

interface MyObject {
  [key: string]: any
}

export type ChildType<T extends MyObject> = DefineComponent<{
  value: PropType<T>,
  key: PropType<keyof T>,
}>;
</script>
```

```vue
// parent.vue
<script lang="ts" setup>
import Child from './child.vue'
import type { ChildType } from './child.vue'

const ChildOne = Child as any as ChildType<{ foo: string }>;
const ChildTwo = Child as any as ChildType<{ bar: number }>;
</script>
```

**Reply by sheremet-va** [reply]
Thank you very much! For your answers and for your awesome work. 👍👏

**Comment by sheremet-va** [comment]
Good day, Johnson! ! It's me again! (haha)

This approach works really well with props, but I lost all my slot's types. Can I fix this somehow? And also include generics inside them? Or is this a Volar specific feature and there is nothing I can do?

**Reply by johnsoncodehk** [reply]
Volar define template slots types to `__VLS_slots`, so you can steal this type.

```ts
const ChildOne = Child as any as ChildType<{ foo: string }> & { __VLS_slots: typeof Child['__VLS_slots'] };
```

If with generics, you need to define it with `ChileType`.

```ts
export type ChildType<T extends MyObject> = DefineComponent<{
    value: PropType<T>,
    key: PropType<keyof T>,
}> & {
    __VLS_slots: {
        mySlot: {
            slotValue: T,
            slotKey: keyof T,
        }
    }
};
```

How your problems funny. 🤣

**Reply by sheremet-va** [reply]
> How your problems funny. 🤣

Yeah, experiments, experiments! Just trying to make the most type safe component 🌝 

> If with generics, you need to define it with ChileType.

Thank you! Works very well.

**Reply by Evertt** [reply]
@johnsoncodehk your code brought me so close to what I need and I'm super grateful. Would you be willing to see if you can help me the last 5%?

Right now I have this:

```ts
export type VTablesType<T extends () => Api> = DefineComponent<
  Modify<Props, {api: T}>
> & {
  __VLS_slots: {
    reference: { row: UnwrapRef<ReturnType<T>["data"]> }
  }
}
```

And it's allowing me to use a slot like this:

```html
<template #reference="{row}">
    <span>{{ row.reference }}</span>
</template>
```

And `row` is correctly typed there. Now the thing is, in my particular case, I know that ALL slots will give the same data type. And I have A LOT of slots. So many that it would be super tedious to add them all separately to `__VLS_slots`. So I hoped I could fix that by writing it like so:

```ts
__VLS_slots: {
   [key: string]: { row: UnwrapRef<ReturnType<T>["data"]> }
}
```

But unfortunately that doesn't work, then every slot says their data is `any` again...
Do you have any idea how I could achieve this? I would be so grateful 🙏

**Reply by johnsoncodehk** [reply]
@Evertt this comment may helps: https://github.com/johnsoncodehk/volar/issues/877#issuecomment-1060127400

Please note that with #1108, we will not support `__VLS_slots` any more, please use `$slots` instead of.

**Comment by johnsoncodehk** [comment]
@sheremet-va cc: https://github.com/vuejs/rfcs/pull/310

**Comment by johnsoncodehk** [comment]
Generic functional component already supported in v0.25.2.

```vue
// child.vue
<script lang="ts">
import { h } from "vue";

const Comp = <T extends object>(props: {
  items: T;
  keys?: keyof T;
}) => h({}, props);

export default Comp;
</script>
```

```vue
<template>
  <Child :items="{ foo: '' }" keys="foo" />
  <Child :items="{ foo: '' }" keys="bar" /> <!-- Type '"bar"' is not assignable to type '"foo" | undefined' -->
</template>

<script lang="ts" setup>
import Child from './child.vue'
</script>
```

**Reply by sheremet-va** [reply]
It’s sad I don’t use those, but thank you for heads up

**Comment by rndmerle** [comment]
Thanks for the help.
My not-pretty temporary fix for our Vue 2 project is this for now:
```
components: {
    // to get v-slot="{ item }" typed
    MyGenericComponent: MyGenericComponent as typeof MyGenericComponent & {
      __VLS_slots: {
        default: { item: ItemType }
      }
    }
```
